app/models/bag.rb

```diff
@@ -0,0 +1,44 @@
+# Creates the structure for an individual thesis to be preserved in Archivematica according to the BagIt spec:
+# https://datatracker.ietf.org/doc/html/rfc8493.
+class Bag
+  include Checksums
+
+  def initialize(thesis)
+    @thesis = thesis
+    raise 'This thesis is not baggable' unless baggable?
+  end
+
+  def data
+    file_locations = {}
+    @thesis.files.map { |f| file_locations["data/#{f.filename}"] = f.blob }
+
+    # TODO: build csv, add it to hash
+
+    file_locations
+  end
+
+  def bag_declaration
+    "BagIt-Version: 1.0\nTag-File-Character-Encoding: UTF-8"
+  end
+
+  def manifest
+    @thesis.files.map { |f| "#{base64_to_hex(f.checksum)} data/#{f.filename}" }.join("\n")
+  end
+
+  def bag_name
+    safe_handle = @thesis.dspace_handle.gsub('/', '_')
+    "#{safe_handle}-thesis"
+  end
+
+  # Before we try to bag anything, we need to check if it meets a few conditions. All published theses should have
+  # at least one file attached, no duplicate filenames, and a handle pointing to its DSpace record.
+  # TODO: add a check for metadata csv
+  def baggable?
+    @thesis.files.any? && @thesis.dspace_handle && !duplicate_filenames?
```

Comment:
I've noticed an edge case we may need to cover. Locally I had a thesis in which I had manually removed a handle in the UI. The result ends up being `dspace_handle` is an empty string rather than `nil` which this logic considers baggable. I'm not sure how frequently this might come up in production, but I think I'd like to handle it. (ha!). It may be a simple as changing the check to `@thesis.dspace_handle.blank?` which I believe accounts for empty strings, strings with only whitespace, and nil... maybe

Reply:
Good call. Yeah, I think `blank?` should do it. Thanks!